File: scheduler/tasks.py
# -*- coding: utf-8 -*-
"""
Модуль планировщика задач — сбор, фильтрация и публикация вакансий.

Содержит основную задачу daily_job(), которая:
1. Собирает вакансии из всех включённых источников
2. Фильтрует по заданным критериям
3. Проверяет дубликаты через базу данных
4. Публикует новые вакансии в Telegram-канал
"""

import sys
import os
import hashlib
import asyncio
from typing import List, Dict, Any

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from config import SOURCES, MAX_PER_SESSION, DELAY_BETWEEN_POSTS
from db.database import Database
from filters.vacancy_filter import filter_vacancies
from templates.post_template import format_post, format_post_short
from bot.publisher import Publisher
from loguru import logger


async def _fetch_from_source(source_name: str) -> List[Dict[str, Any]]:
    """
    Получает вакансии из указанного источника.

    Args:
        source_name: имя источника (hh, trudvsem, superjob, tg, vk, habr, avito)

    Returns:
        Список словарей с данными вакансий.
    """
    vacancies: List[Dict[str, Any]] = []

    try:
        if source_name == "hh":
            from sources.hh_api import fetch_hh_vacancies
            vacancies = await fetch_hh_vacancies()

        elif source_name == "trudvsem":
            from sources.trudvsem_api import fetch_trudvsem_vacancies
            vacancies = await fetch_trudvsem_vacancies()

        elif source_name == "superjob":
            from sources.superjob_api import fetch_superjob_vacancies
            vacancies = await fetch_superjob_vacancies()

        elif source_name == "tg":
            from sources.tg_parser import fetch_tg_vacancies
            vacancies = await fetch_tg_vacancies()

        elif source_name == "vk":
            from sources.vk_parser import fetch_vk_vacancies
            vacancies = await fetch_vk_vacancies()

        elif source_name == "habr":
            from sources.habr_parser import fetch_habr_vacancies
            vacancies = await fetch_habr_vacancies()

        elif source_name == "avito":
            from sources.avito_parser import fetch_avito_vacancies
            vacancies = await fetch_avito_vacancies()

        else:
            logger.warning(f"Неизвестный источник: {source_name}")
            return []

        logger.info(f"Источник '{source_name}': получено {len(vacancies)} вакансий")

    except ImportError as e:
        logger.error(f"Не удалось импортировать модуль для источника '{source_name}': {e}")
    except Exception as e:
        logger.error(f"Ошибка при сборе вакансий из '{source_name}': {e}")

    return vacancies


def _title_hash(title: str) -> str:
    """
    Вычисляет SHA-256 хеш от заголовка вакансии (в нижнем регистре, без пробелов по краям).

    Args:
        title: заголовок вакансии

    Returns:
        Hex-строка SHA-256 хеша.
    """
    normalized = title.strip().lower()
    return hashlib.sha256(normalized.encode("utf-8")).hexdigest()


async def daily_job() -> None:
    """
    Основная задача планировщика — полный цикл сбора и публикации вакансий.

    Последовательность:
    1. Сбор вакансий из всех включённых источников (SOURCES)
    2. Фильтрация по ключевым словам / стоп-словам / зарплате
    3. Проверка на дубликаты через БД (по source+id, url, title_hash)
    4. Форматирование и публикация в Telegram-канал (до MAX_PER_SESSION штук)
    5. Сохранение опубликованных записей в БД
    """
    logger.info("=" * 60)
    logger.info("Запуск задачи сбора и публикации вакансий")
    logger.info("=" * 60)

    db = Database()
    publisher = Publisher()

    try:
        await db.init_db()

        # ── 1. Сбор вакансий из всех включённых источников ──────────
        all_vacancies: List[Dict[str, Any]] = []

        for source_name, enabled in SOURCES.items():
            if not enabled:
                logger.debug(f"Источник '{source_name}' отключён, пропускаем")
                continue

            fetched = await _fetch_from_source(source_name)
            all_vacancies.extend(fetched)

        logger.info(f"Всего собрано вакансий (до фильтрации): {len(all_vacancies)}")

        if not all_vacancies:
            logger.warning("Не удалось собрать ни одной вакансии, завершаем")
            return

        # ── 2. Фильтрация ──────────────────────────────────────────
        filtered = filter_vacancies(all_vacancies)
        logger.info(f"После фильтрации: {len(filtered)} вакансий")

        if not filtered:
            logger.info("После фильтрации не осталось вакансий")
            return

        # ── 3. Проверка дубликатов ──────────────────────────────────
        new_vacancies: List[Dict[str, Any]] = []

        for vacancy in filtered:
            source = vacancy.get("source", "unknown")
            vacancy_id = vacancy.get("vacancy_id", "")
            url = vacancy.get("url", "")
            title = vacancy.get("title", "")
            company = vacancy.get("company", "")

            # Проверяем по source + vacancy_id
            if vacancy_id and await db.exists(source, str(vacancy_id)):
                logger.debug(f"Дубликат (source+id): {title[:50]}")
                continue

            # Проверяем по URL
            if url and await db.exists_by_url(url):
                logger.debug(f"Дубликат (url): {title[:50]}")
                continue

            # Проверяем по хешу заголовка
            if title and company and await db.exists_by_hash(title, company):
                logger.debug(f"Дубликат (title_hash): {title[:50]}")
                continue

            new_vacancies.append(vacancy)

        logger.info(f"Новых уникальных вакансий: {len(new_vacancies)}")

        if not new_vacancies:
            logger.info("Нет новых вакансий для публикации")
            return

        # ── 4. Публикация ───────────────────────────────────────────
        published_count = 0

        for vacancy in new_vacancies[:MAX_PER_SESSION]:
            try:
                title = vacancy.get("title", "")
                description = vacancy.get("description", "")

                # Выбираем шаблон в зависимости от наличия описания
                if description and description.strip():
                    text = format_post(vacancy)
                else:
                    text = format_post_short(vacancy)

                success = await publisher.send_to_channel(text)

                if success:
                    # Сохраняем в БД
                    await db.insert(vacancy)
                    published_count += 1
                    logger.info(
                        f"[{published_count}/{MAX_PER_SESSION}] "
                        f"Опубликовано: {title[:60]}"
                    )
                else:
                    logger.warning(f"Не удалось опубликовать: {title[:60]}")

            except Exception as e:
                logger.error(
                    f"Ошибка при публикации вакансии "
                    f"'{vacancy.get('title', '?')[:40]}': {e}"
                )

            # Задержка между постами, чтобы не нарваться на лимиты Telegram
            if published_count < MAX_PER_SESSION:
                await asyncio.sleep(DELAY_BETWEEN_POSTS)

        logger.info(f"Итого опубликовано: {published_count} вакансий")

    except Exception as e:
        logger.exception(f"Критическая ошибка в daily_job: {e}")

    finally:
        await publisher.close()
        await db.close()
        logger.info("Задача завершена, ресурсы освобождены")


async def test_source(source_name: str) -> None:
    """
    Тестовый запуск одного источника — загружает вакансии и выводит в лог
    без публикации в канал. Используется для отладки.

    Args:
        source_name: имя источника (hh, trudvsem, superjob, tg, vk, habr, avito)
    """
    logger.info(f"Тестовый запуск источника: {source_name}")

    vacancies = await _fetch_from_source(source_name)

    if not vacancies:
        logger.warning(f"Источник '{source_name}' не вернул вакансий")
        return

    logger.info(f"Получено {len(vacancies)} вакансий из '{source_name}':")
    logger.info("-" * 60)

    for i, v in enumerate(vacancies, start=1):
        title = v.get("title", "—")
        company = v.get("company", "—")
        sal_from = v.get("salary_from")
        sal_to = v.get("salary_to")
        salary = f"{sal_from or ''} - {sal_to or ''}".strip(" -") or "не указана"
        area = v.get("area", "—")
        url = v.get("url", "—")

        logger.info(
            f"  [{i}] {title}\n"
            f"       Компания: {company}\n"
            f"       Зарплата: {salary}\n"
            f"       Регион:   {area}\n"
            f"       URL:      {url}"
        )

    logger.info("-" * 60)
    logger.info(f"Всего: {len(vacancies)}")

    # Показываем результат фильтрации (но не публикуем)
    filtered = filter_vacancies(vacancies)
    logger.info(f"После фильтрации осталось: {len(filtered)} вакансий")

File: src/scheduler/tasks.ts
import { config } from '../config/config.js';
import { Database, Vacancy } from '../db/database.js';
import { logger } from '../utils/logger.js';
import { Publisher } from '../bot/publisher.js';
import { filterVacancies } from '../filters/vacancyFilter.js';
import { formatPost } from '../templates/postTemplate.js';
import {
  fetchHhVacancies,
  fetchTrudvsemVacancies,
  fetchSuperjobVacancies,
  fetchHabrVacancies,
  fetchVkVacancies,
  fetchTgVacancies,
  fetchAvitoVacancies
} from '../sources/index.js';

// Helper to sleep/delay execution
const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

export async function dailyJob() {
  logger.info('Daily Job: Starting cycle...');
  
  const db = new Database();
  const publisher = new Publisher();
  
  let rawVacancies: Vacancy[] = [];

  // Run enabled sources concurrently using Promise.allSettled
  const activeSources: { name: string; promise: Promise<Vacancy[]> }[] = [];

  if (config.SOURCES.hh) activeSources.push({ name: 'hh', promise: fetchHhVacancies() });
  if (config.SOURCES.trudvsem) activeSources.push({ name: 'trudvsem', promise: fetchTrudvsemVacancies() });
  if (config.SOURCES.superjob) activeSources.push({ name: 'superjob', promise: fetchSuperjobVacancies() });
  if (config.SOURCES.habr) activeSources.push({ name: 'habr', promise: fetchHabrVacancies() });
  if (config.SOURCES.vk) activeSources.push({ name: 'vk', promise: fetchVkVacancies() });
  if (config.SOURCES.tg) activeSources.push({ name: 'tg', promise: fetchTgVacancies() });
  if (config.SOURCES.avito) activeSources.push({ name: 'avito', promise: fetchAvitoVacancies() });

  logger.info(`Daily Job: Triggering ${activeSources.length} active source parsers...`);

  const results = await Promise.allSettled(activeSources.map(s => s.promise));

  for (let i = 0; i < results.length; i++) {
    const res = results[i];
    const sourceName = activeSources[i].name;

    if (res.status === 'fulfilled') {
      rawVacancies = rawVacancies.concat(res.value);
    } else {
      logger.error(`Daily Job: Source "${sourceName}" failed to fetch: ${res.reason?.message || res.reason}`);
    }
  }

  logger.info(`Daily Job: Collected ${rawVacancies.length} raw vacancies from all sources.`);

  // 1. Filter out vacancies (salary, remote, stop-words)
  const filtered = filterVacancies(rawVacancies);
  logger.info(`Daily Job: ${filtered.length} vacancies left after applying filters.`);

  // 2. Filter out duplicates (anti-duplication)
  const newVacancies: Vacancy[] = [];
  for (const vacancy of filtered) {
    // Check by source + id
    if (db.exists(vacancy.source, vacancy.vacancy_id)) {
      logger.debug(`Deduplication: Skipping already published ID: ${vacancy.source}:${vacancy.vacancy_id}`);
      continue;
    }

    // Check by url
    if (vacancy.url && db.existsByUrl(vacancy.url)) {
      logger.debug(`Deduplication: Skipping already published URL: ${vacancy.url}`);
      continue;
    }

    // Check by title + company hash
    if (vacancy.title && vacancy.company && db.existsByHash(vacancy.title, vacancy.company)) {
      logger.debug(`Deduplication: Skipping duplicate hash for "${vacancy.title}" @ "${vacancy.company}"`);
      continue;
    }

    newVacancies.push(vacancy);
  }

  logger.info(`Daily Job: Found ${newVacancies.length} brand new vacancies to process.`);

  // 3. Limit to MAX_PER_SESSION
  const targetBatch = newVacancies.slice(0, config.MAX_PER_SESSION);
  logger.info(`Daily Job: Publishing up to ${targetBatch.length} vacancies in this session (limit: ${config.MAX_PER_SESSION}).`);

  let publishedCount = 0;

  // 4. Publish vacancies sequentially with delay
  for (const vacancy of targetBatch) {
    const text = formatPost(vacancy);
    
    // Publish
    const success = await publisher.sendToChannel(text);
    if (success) {
      // Save to database
      db.insert(vacancy);
      publishedCount++;
      
      // Delay to avoid hitting flood limits
      if (publishedCount < targetBatch.length) {
        logger.debug(`Sleeping for ${config.DELAY_BETWEEN_POSTS} seconds before next post...`);
        await sleep(config.DELAY_BETWEEN_POSTS * 1000);
      }
    }
  }

  logger.info(`Daily Job: Cycle complete. Published ${publishedCount}/${targetBatch.length} vacancies.`);

  // Cleanup connections
  db.close();
  await publisher.close();
}

export async function testSource(sourceName: string) {
  logger.info(`Test Mode: Testing single source parser "${sourceName}"...`);
  
  let fetcher: () => Promise<Vacancy[]>;

  switch (sourceName.toLowerCase()) {
    case 'hh':
      fetcher = fetchHhVacancies;
      break;
    case 'trudvsem':
      fetcher = fetchTrudvsemVacancies;
      break;
    case 'superjob':
      fetcher = fetchSuperjobVacancies;
      break;
    case 'habr':
      fetcher = fetchHabrVacancies;
      break;
    case 'vk':
      fetcher = fetchVkVacancies;
      break;
    case 'tg':
      fetcher = fetchTgVacancies;
      break;
    case 'avito':
      fetcher = fetchAvitoVacancies;
      break;
    default:
      logger.error(`Unknown source parser name: ${sourceName}. Available options: hh, trudvsem, superjob, habr, vk, tg, avito`);
      return;
  }

  try {
    const fetched = await fetcher();
    logger.info(`Test Mode: Found ${fetched.length} raw vacancies.`);

    const filtered = filterVacancies(fetched);
    logger.info(`Test Mode: ${filtered.length} passed filter criteria.`);

    if (filtered.length > 0) {
      logger.info('--- SAMPLE FORMATED POST (first 2 items) ---');
      const sampleBatch = filtered.slice(0, 2);
      for (const item of sampleBatch) {
        console.log('\n=============================================');
        console.log(formatPost(item));
        console.log('=============================================\n');
      }
    } else {
      logger.warn('Test Mode: No vacancies passed filters to show formatting.');
    }
  } catch (error: any) {
    logger.error(`Test Mode: Failed during test execution of ${sourceName}: ${error.message}`);
  }
}
